### PR TITLE
Normalise `dateModified` format in `query_last_value` function

### PR DESCRIPTION
**Pull request by chicco785**
## Proposed changes

While all other returned by `query_last_value` function where normalised, `dateModified` is not.
This PR Normalise `dateModified` format in `query_last_value` function

## Types of changes

What types of changes does your code introduce to the project: _Put
an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing
      functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after
creating the PR. If you're unsure about any of them, don't hesitate to
ask. We're here to help! This is simply a reminder of what we are going
to look for before merging your code._

- [x] I have read the [CONTRIBUTING][contrib] doc
- [x] I have signed the [CLA][cla]
- [x] I have added tests that prove my fix is effective or that my
      feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in
      downstream modules

## Further comments

N/A

[cla]: https://raw.githubusercontent.com/orchestracities/ngsi-timeseries-api/master/individual-cla.pdf
    "Martel Open Source Software Individual Contributor License Agreement"
[contrib]: https://github.com/smartsdk/ngsi-timeseries-api/blob/master/CONTRIBUTING.md
    "Contributing to QuantumLeap"

### CODE REVIEW

**Comment by chicco785** [comment]
@pooja1pathak I normalised `dateModified` so that we don't need to pop it, and also cleaned up few things. if you could review changes, it's great, thanks!

**Comment by chicco785** [comment]
fix #547 and #548

**Comment by pooja1pathak** [comment]
> @pooja1pathak I normalised `dateModified` so that we don't need to pop it, and also cleaned up few things. if you could review changes, it's great, thanks!

sure

**Review by pooja1pathak — CHANGES_REQUESTED** [review]
we need to change fiware-servicePath value to None for the case '/' is received.
In QL if we directly hit /op/query, it will take fiware-service and fiware-servicePath as None but in ```orion``` it add '/' to the fiware-servicePath which results in formation of statement as:
```
('select '
 'a1.entity_type,a1.entity_id,a1.time_index,a1."temperature",a1."intensity" '
 'from "ettestnew" as a1 join (select entity_id, entity_type, max(time_index) '
 'as time_index from "ettestnew" where entity_id in (\'TestNew1\')  and  '
 "fiware_servicepath ~* '/.*' group by entity_id, entity_type) b1 on "
 'a1.entity_id = b1.entity_id and a1.entity_type = b1.entity_type and '
 "a1.time_index = b1.time_index where a1.entity_id in ('TestNew1')  and  "
 "a1.fiware_servicepath ~* '/.*' ORDER BY time_index DESC limit 10000 offset 0")
```
Notice the value of **fiware_servicepath**. This results in 404 Not Found.

For None fiware_servicePath statement for same request is as:
```
 ('select '
 'a1.entity_type,a1.entity_id,a1.time_index,a1."temperature",a1."intensity" '
 'from "ettestnew" as a1 join (select entity_id, entity_type, max(time_index) '
 'as time_index from "ettestnew" where entity_id in (\'TestNew1\')  and  '
 "fiware_servicepath = '' group by entity_id, entity_type) b1 on a1.entity_id "
 '= b1.entity_id and a1.entity_type = b1.entity_type and a1.time_index = '
 "b1.time_index where a1.entity_id in ('TestNew1')  and  a1.fiware_servicepath "
 "= '' ORDER BY time_index DESC limit 10000 offset 0")
```

**Comment by chicco785** [comment]
> we need to change fiware-servicePath value to None for the case '/' is received.
> In QL if we directly hit /op/query, it will take fiware-service and fiware-servicePath as None but in `orion` it add '/' to the fiware-servicePath which results in formation of statement as:
> 
> ```
> ('select '
>  'a1.entity_type,a1.entity_id,a1.time_index,a1."temperature",a1."intensity" '
>  'from "ettestnew" as a1 join (select entity_id, entity_type, max(time_index) '
>  'as time_index from "ettestnew" where entity_id in (\'TestNew1\')  and  '
>  "fiware_servicepath ~* '/.*' group by entity_id, entity_type) b1 on "
>  'a1.entity_id = b1.entity_id and a1.entity_type = b1.entity_type and '
>  "a1.time_index = b1.time_index where a1.entity_id in ('TestNew1')  and  "
>  "a1.fiware_servicepath ~* '/.*' ORDER BY time_index DESC limit 10000 offset 0")
> ```
> 
> Notice the value of **fiware_servicepath**. This results in 404 Not Found.
> 
> For None fiware_servicePath statement for same request is as:
> 
> ```
>  ('select '
>  'a1.entity_type,a1.entity_id,a1.time_index,a1."temperature",a1."intensity" '
>  'from "ettestnew" as a1 join (select entity_id, entity_type, max(time_index) '
>  'as time_index from "ettestnew" where entity_id in (\'TestNew1\')  and  '
>  "fiware_servicepath = '' group by entity_id, entity_type) b1 on a1.entity_id "
>  '= b1.entity_id and a1.entity_type = b1.entity_type and a1.time_index = '
>  "b1.time_index where a1.entity_id in ('TestNew1')  and  a1.fiware_servicepath "
>  "= '' ORDER BY time_index DESC limit 10000 offset 0")
> ```

hi @pooja1pathak i am not sure i understand what's the actual issue :/ i will add a test case using `/` servicePath and see what's the query outcome. would that allow us to understand if the code works?

File: src/reporter/tests/test_op.py [hunk 1]
```diff
@@ -217,9 +256,76 @@ def test_query_no_id(service, reporter_dataset):
...
-    assert r.json() == {
-        'detail': "'id' is a required property - 'entities.0'",
-        'status': 400,
-        'title': 'Bad Request',
-        'type': 'about:blank'
+    assert r.json() == "Entity id is required"
+
+
+@pytest.mark.parametrize("service", services)
+def test_default_service_path(service):
```
Comment (chicco785) [inline src/reporter/tests/test_op.py:263]: @pooja1pathak first newly introduced test

File: src/reporter/tests/test_op.py [hunk 2]
```diff
@@ -217,9 +256,76 @@ def test_query_no_id(service, reporter_dataset):
...
+                      headers=headers(service, alt_service_path))
+    assert r.status_code == 200, r.text
+    assert len(r.json()) == 1
+    assert r.json()[0]['temperature']['value'] == 14
+    delete_test_data(service, [entity_type], service_path=service_path)
+    delete_test_data(service, [entity_type], service_path=alt_service_path)
+
+
+@pytest.mark.parametrize("service", services)
+def test_none_service_path(service):
```
Comment (chicco785) [inline src/reporter/tests/test_op.py:299]: @pooja1pathak second newly introduced test

**Comment by chicco785** [comment]
> > we need to change fiware-servicePath value to None for the case '/' is received.
> > In QL if we directly hit /op/query, it will take fiware-service and fiware-servicePath as None but in `orion` it add '/' to the fiware-servicePath which results in formation of statement as:
> > ```
> > ('select '
> >  'a1.entity_type,a1.entity_id,a1.time_index,a1."temperature",a1."intensity" '
> >  'from "ettestnew" as a1 join (select entity_id, entity_type, max(time_index) '
> >  'as time_index from "ettestnew" where entity_id in (\'TestNew1\')  and  '
> >  "fiware_servicepath ~* '/.*' group by entity_id, entity_type) b1 on "
> >  'a1.entity_id = b1.entity_id and a1.entity_type = b1.entity_type and '
> >  "a1.time_index = b1.time_index where a1.entity_id in ('TestNew1')  and  "
> >  "a1.fiware_servicepath ~* '/.*' ORDER BY time_index DESC limit 10000 offset 0")
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Notice the value of **fiware_servicepath**. This results in 404 Not Found.
> > For None fiware_servicePath statement for same request is as:
> > ```
> >  ('select '
> >  'a1.entity_type,a1.entity_id,a1.time_index,a1."temperature",a1."intensity" '
> >  'from "ettestnew" as a1 join (select entity_id, entity_type, max(time_index) '
> >  'as time_index from "ettestnew" where entity_id in (\'TestNew1\')  and  '
> >  "fiware_servicepath = '' group by entity_id, entity_type) b1 on a1.entity_id "
> >  '= b1.entity_id and a1.entity_type = b1.entity_type and a1.time_index = '
> >  "b1.time_index where a1.entity_id in ('TestNew1')  and  a1.fiware_servicepath "
> >  "= '' ORDER BY time_index DESC limit 10000 offset 0")
> > ```
> 
> hi @pooja1pathak i am not sure i understand what's the actual issue :/ i will add a test case using `/` servicePath and see what's the query outcome. would that allow us to understand if the code works?

I added two new tests, aiming at surfacing any potential issues related to using `/` or `None`, but it seems to me that both works fine.

**Comment by pooja1pathak** [comment]
@chicco785 I have checked, test cases seems to work fine. Issue will arise when only ```fiware-servicePath``` is mentioned i.e., without any ```fiware-service```. 

As I have checked in crateDB, ```fiware-servicePath``` is stored only when ```fiware-service``` is mentioned otherwise it is left empty whether we provide any ```fiware-servicePath``` or not. As per my understanding this could be the reason for the 404 error.

I have checked with ```orion``` broker also by creating entity with ```fiware-servicePath``` only without defining any ```fiware-service``` . I was able to retrieve entity with as well as without using ```fiware-servicePath```. I think this should not be the correct behavior.

**Comment by chicco785** [comment]
> @chicco785 I have checked, test cases seems to work fine. Issue will arise when only `fiware-servicePath` is mentioned i.e., without any `fiware-service`.
> 
> As I have checked in crateDB, `fiware-servicePath` is stored only when `fiware-service` is mentioned otherwise it is left empty whether we provide any `fiware-servicePath` or not. As per my understanding this could be the reason for the 404 error.
> 
> I have checked with `orion` broker also by creating entity with `fiware-servicePath` only without defining any `fiware-service` . I was able to retrieve entity with as well as without using `fiware-servicePath`. I think this should not be the correct behavior.

thanks @pooja1pathak this information is very useful, i will add tests, and let's see the outcomes. this may be quite a complex corner case... fingers crossed!

File: src/wq/ql/notify.py [hunk 3]
```diff
@@ -79,20 +79,12 @@ def retry_intervals(self) -> [int]:
...
         data = self.task_input()
         svc = data.fiware_service
-        svc_path = data.fiware_service_path if svc else None    # (*)
+        svc_path = data.fiware_service_path
         try:
             with translator_for(svc) as trans:
                 trans.insert(data.payload, svc, svc_path)
         except Exception as e:
             self._handle_exception(svc, e)
-# NOTE. Service path.
```
Comment (chicco785) [inline src/wq/ql/notify.py:88]: @c0c0n3 i have no memory of why it was like this... this anyhow seems to introduce a bug when service path is not null, and service is... because whatever data sent will be cast to a null servicePath

Comment (chicco785) [reply src/wq/ql/notify.py:88]: well, clearly given all the mighty bombing we have now on the reporter tests, highlights how this choice was "core" in the implementation, despite leading to uncorrect support of service_paths :/

Comment (c0c0n3) [reply src/wq/ql/notify.py:88]: oh dear, it looks like I've actually introduced a bug! the code was meant to be like for like but obviously it isn't! I must've missed a statement when I copy-pasted the code. It was meant to be exactly the same as it was in the reporter before implementing wq...

Comment (c0c0n3) [reply src/wq/ql/notify.py:88]: well, at least that was my intention judging by the comment in the code...

Comment (chicco785) [reply src/wq/ql/notify.py:88]: i am not sure this was introduced by you :) the note was saying that the logic was taken from the original notify, so i assume that's true, so the question is why it was introduced originally. of course, given that we basically always test against a tenant, we don't observe any problem. testing with tenant `None` the are problems, which are solved by removing, the cast to `None` of the path, but the other comes up...

Comment (chicco785) [reply src/wq/ql/notify.py:88]: so, i confirm it was there:
https://github.com/orchestracities/ngsi-timeseries-api/blob/45aff7a8a94ed4ea67e3cb6ccb93fa06fd7c6b5f/src/reporter/reporter.py#L179

and that is was introduced with multitenancy support 3 years ago:
https://github.com/orchestracities/ngsi-timeseries-api/pull/65/files

Comment (c0c0n3) [reply src/wq/ql/notify.py:88]: oh my! it looks like it;'s a catch 22... :-)

File: src/reporter/tests/utils.py [hunk 4]
```diff
@@ -80,9 +83,9 @@ def insert_test_data(service, entity_types, n_entities=1, index_size=30,
 
                 eid = entity_id or '{}{}'.format(et, ei)
                 n = get_notification(et, eid, attr_value=i, mod_value=dt)
-                send_notifications(service, [n])
-
-    time.sleep(1)
+                send_notifications(service, [n], service_path)
+    # TODO do we need this?
+    # time.sleep(1)
```
Comment (chicco785) [inline src/reporter/tests/utils.py:88]: @c0c0n3 this may be the actual cause of the bombing...

**Comment by chicco785** [comment]
> thanks @pooja1pathak this information is very useful, i will add tests, and let's see the outcomes. this may be quite a complex corner case... fingers crossed!

issue fixed, hopefully

**Comment by pooja1pathak** [comment]
@chicco785 I have verified the changes and found one more issue. In ```orion``` if no header is used then it uses ```fiware-servicePath: /``` but in QL if no header is used then it is set to ```None```. 

I checked this by integrating with orion and I was unable to retrieve entity as it sends ```fiware-servicePath: /``` in Header while sending POST request to QL which results in ```404 Not Found``` error.

If possible, We can raise separate issue to handle this.

**Comment by chicco785** [comment]
> @chicco785 I have verified the changes and found one more issue. In `orion` if no header is used then it uses `fiware-servicePath: /` but in QL if no header is used then it is set to `None`.
> 
> I checked this by integrating with orion and I was unable to retrieve entity as it sends `fiware-servicePath: /` in Header while sending POST request to QL which results in `404 Not Found` error.
> 
> If possible, We can raise separate issue to handle this.

which one of these scenarios did you test?
1. send data with no `fiware-servicePath` to orion and query ql using `fiware-servicePath: /`
1. send data with no `fiware-servicePath` to orion and query ql using no `fiware-servicePath`

**Comment by pooja1pathak** [comment]
> which one of these scenarios did you test?
> 
> 1. send data with no `fiware-servicePath` to orion and query ql using `fiware-servicePath: /`
> 2. send data with no `fiware-servicePath` to orion and query ql using no `fiware-servicePath`

@chicco785 I have tested following scenarios:
1. send data with no fiware-servicePath to orion and query ql using fiware-servicePath: / : Result was ```200 OK```
2. send data with no fiware-servicePath to orion and query ql using no fiware-servicePath : Result was ```404 Not Found```
3. without using fiware-servicePath : Created entity at ql using notify API and register ql in orion as datasource for that entity. Then query orion for entity : Result was ```404 Not Found```

**Comment by chicco785** [comment]
> > which one of these scenarios did you test?
> > 
> > 1. send data with no `fiware-servicePath` to orion and query ql using `fiware-servicePath: /`
> > 2. send data with no `fiware-servicePath` to orion and query ql using no `fiware-servicePath`
> 
> @chicco785 I have tested following scenarios:
> 
> 1. send data with no fiware-servicePath to orion and query ql using fiware-servicePath: / : Result was `200 OK`
> 2. send data with no fiware-servicePath to orion and query ql using no fiware-servicePath : Result was `404 Not Found`
> 3. without using fiware-servicePath : Created entity at ql using notify API and register ql in orion as datasource for that entity. Then query orion for entity : Result was `404 Not Found`

2. and 3. from my point of view are not so much a problem. the solution (and the problem is that i won't be retroactive - unless we create a script to update all tables), would be to always add "/" when servicePath is none. definitively, anyhow, something not to be covered in this PR.

**Review by pooja1pathak — APPROVED** [review]

**Comment by pooja1pathak** [comment]
> > > which one of these scenarios did you test?
> > > 
> > > 1. send data with no `fiware-servicePath` to orion and query ql using `fiware-servicePath: /`
> > > 2. send data with no `fiware-servicePath` to orion and query ql using no `fiware-servicePath`
> > 
> > 
> > @chicco785 I have tested following scenarios:
> > 
> > 1. send data with no fiware-servicePath to orion and query ql using fiware-servicePath: / : Result was `200 OK`
> > 2. send data with no fiware-servicePath to orion and query ql using no fiware-servicePath : Result was `404 Not Found`
> > 3. without using fiware-servicePath : Created entity at ql using notify API and register ql in orion as datasource for that entity. Then query orion for entity : Result was `404 Not Found`
> 
> 2. and 3. from my point of view are not so much a problem. the solution (and the problem is that i won't be retroactive - unless we create a script to update all tables), would be to always add "/" when servicePath is none. definitively, anyhow, something not to be covered in this PR.

Yes, We can raise separate issue and PR to handle this scenario.
